Add Translation (a TransformComponent)

nattridgell opened the pull request:
Key points for a translation:
- constructor
  - accepts 2 or 3 LengthValues as arguments
  - currently only SimpleLengths with type `px` are supported
    (This is due to other types needing to be resolved.)
- z is null if it is a 2D component
- matrix is an identity matrix except for the last column
  - has [x, y] as the last column for 2D
  - has [x, y, z, 1] as the last column for 3D
  - where x, y, z are the px lengths of the length values
- cssString
  - either translate / translate3d
  - has a comma seperated list of the length value cssStrings

Commit:
Add Translation (a TransformComponent)

Key points for a translation:
 - constructor
   - accepts 2 or 3 LengthValues as arguments
   - currently only SimpleLengths with type `px` are supported
     (This is due to other types needing to be resolved.)
 - z is null if it is a 2D component
 - matrix is an identity matrix, except for the last column
   - has [x, y] as the last column for 2D
   - has [x, y, z, 1] as the last column for 3D
   - where x, y, z are the px lengths of the length values
 - cssString
   - either translate / translate3d
   - has a comma seperated list of the length value cssStrings

wilddamon commented on src/translation.js:
```diff
@@ -0,0 +1,70 @@
+      } else if (!(arguments[i] instanceof SimpleLength) ||
+          arguments[i].type != 'px') {
+        throw new TypeError('Unsupported LengthValue for Translation. ' +
+            'Only SimpleLength instances with type \'px\' are supported.');
```
Can you just skip the first check then if they have to be SimpleLengths anyway?

wilddamon commented on test/js/translation.js:
```diff
@@ -0,0 +1,80 @@
+    assert.throws(function() {new Translation(validLength, pxCalcLength)});
+    assert.throws(function() {new Translation(pxCalcLength, pxCalcLength)});
+    assert.throws(function() {new Translation(validLength)});
+    assert.throws(function() {new Translation(validLength, validLength, null)});
```
maybe add some comments for why these are invalid? (the last two are a little confusing since they only contain the variable "validLength")

Commit:
Remove check for LengthValue type in Translation

 - only check that each argument is a SimpleLength with type `px`
 - explain what some of the tests are checking.

Commit:
Initialise all values in Translation constructor

 - alter _computeMatrix and _generateCssString
   to return the matrix / cssString
 - initialise _matrix and cssString in constructor

Commit:
Style update

Run gjslint on translation files.

rjwright commented on src/translation.js:
```diff
@@ -0,0 +1,72 @@
+  };
+
+  Translation.prototype._computeMatrix = function() {
+    // Translation represented by the identity matrix with the translation
```
I'd change this to 'Translation is represented', just for clarity

rjwright commented on test/js/translation.js:
```diff
@@ -0,0 +1,82 @@
+  test('Translation is a Translation and TransformComponent', function() {
+    var validLength = new SimpleLength(1, 'px');
+    var translation = new Translation(validLength, validLength);
+    assert.instanceOf(translation, Translation,
```
No need for this. You just constructed it so you know this will pass :)

Commit:
Remove unnecessary check in test

Remove check that a translation is a translation.
(Unnecessary.)

Improve commenting.

wilddamon commented on test/js/translation.js:
```diff
@@ -0,0 +1,80 @@
+    assert.throws(function() {new Translation()});
+    assert.throws(function() {new Translation({})});
+    assert.throws(function() {new Translation('1px', '2px')});
+    assert.throws(function() {new Translation(null, null)});
```
What about a check for an invalid third argument, here or another test method?

wilddamon commented:
LGTM other than comment about tests

Commit:
Translation Constructor Tests (Expand)

Add more tests for different constructors.

Ensure that if the z argument is invalid, that a 2D translation
isn't made.

wilddamon commented:
LGTM
